sky/server/common.py

```diff
@@ -0,0 +1,441 @@
+
+
+def handle_request_error(response: requests.Response) -> None:
+    if response.status_code != 200:
+        with ux_utils.print_exception_no_traceback():
+            raise RuntimeError(
+                'Failed to process response from SkyPilot API server at '
+                f'{get_server_url()}. '
+                f'Response: {response.status_code} '
+                f'{response.text}')
+
+
+def get_request_id(response: requests.Response) -> RequestId:
+    handle_request_error(response)
+    request_id = response.headers.get('X-Request-ID')
+    if request_id is None:
+        with ux_utils.print_exception_no_traceback():
+            raise RuntimeError(
+                'Failed to get request ID from SkyPilot API server at '
+                f'{get_server_url()}. Response: {response.status_code} '
+                f'{response.text}')
+    return request_id
+
+
+def _start_api_server(deploy: bool = False, host: str = '127.0.0.1'):
+    """Starts a SkyPilot API server locally."""
+    # Lock to prevent multiple processes from starting the server at the
+    # same time, causing issues with database initialization.
+    server_url = get_server_url(host)
+    assert server_url in AVAILABLE_LOCAL_API_SERVER_URLS, (
+        f'server url {server_url} is not a local url')
+    with rich_utils.client_status('Starting SkyPilot API server'):
+        logger.info(f'{colorama.Style.DIM}Failed to connect to '
+                    f'SkyPilot API server at {server_url}. '
+                    'Starting a local server.'
+                    f'{colorama.Style.RESET_ALL}')
+        start_uvicorn_in_background(deploy=deploy, host=host)
+        logger.info(ux_utils.finishing_message('SkyPilot API server started.'))
+
+
+def check_server_healthy(endpoint: Optional[str] = None,) -> None:
+    """Check if the API server is healthy.
+
+    Args:
+        endpoint (Optional[str]): The endpoint of the API server.
+        If None, the default endpoint will be used.
+
+    Raises:
+        RuntimeError: If the server is not healthy or the client version does
+            not match the server version.
+    """
+    endpoint = endpoint if endpoint is not None else get_server_url()
+    api_server_info = get_api_server_status(endpoint)
+    api_server_status = api_server_info.status
+    if api_server_status == ApiServerStatus.VERSION_MISMATCH:
+        with ux_utils.print_exception_no_traceback():
+            raise RuntimeError(
+                SKY_API_VERSION_WARNING.format(
+                    server_version=api_server_info.api_version,
+                    client_version=server_constants.API_VERSION))
+    elif api_server_status == ApiServerStatus.UNHEALTHY:
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.ApiServerConnectionError(endpoint)
+
+
+def check_server_healthy_or_start_fn(deploy: bool = False,
+                                     host: str = '127.0.0.1'):
+    try:
+        check_server_healthy()
```

Comment:
I encountered an issue where the server was running but unreachable in https://github.com/skypilot-org/skypilot/pull/4710. Run `sky launch` again will retry creating the server in such case, which overwrites the log file and obscures the real problem. I suggest we extend the creation filelock to a filelock of local server singleton and hint users to troubleshoot or clear up the existing server in such case.